include/math/utils/ndop.h

```diff
@@ -0,0 +1,221 @@
+
+    DOP transformed(const math::float4x4& transform) const noexcept;
+    constexpr DOP unite_with(const DOP& rhv) const noexcept;
+    constexpr DOP unite_with(const math::float3& point) const noexcept;
+
+    constexpr void expand (const math::float3& point) noexcept;
+    
+    constexpr bool contains(const math::float3& point) const noexcept;
+
+   private:
+    std::array<float, get_num_planes()> distances_ 
+    {
+        [](){
+            std::array<float, get_num_planes()> result;
+            std::fill(std::begin(result), std::end(result), std::numeric_limits<float>::lowest());
+            return result;
+        }()
+    };
+};
+
+constexpr std::optional<math::float3> DOP::get_center() const noexcept {
+    return is_empty() ? std::nullopt : std::optional{math::float3{std::midpoint(distances_[0], -distances_[get_inverse_plane_index(0)]),
+                                                      std::midpoint(distances_[1], -distances_[get_inverse_plane_index(1)]),
+                                                      std::midpoint(distances_[2], -distances_[get_inverse_plane_index(2)]),
+    }};
+}
+
+constexpr bool DOP::is_empty() const noexcept { 
+    return std::any_of(std::begin(distances_), std::end(distances_), [](const auto& d) { return d <= std::numeric_limits<float>::lowest(); });
+}
+
+constexpr math::float3 DOP::get_size() const noexcept {
+    return is_empty() ? math::float3{} : math::float3{distances_[0] + distances_[get_inverse_plane_index(0)], distances_[1] + distances_[get_inverse_plane_index(1)], distances_[2] + distances_[get_inverse_plane_index(2)]} ;
+}
+
+template <typename Iterator, typename UnaryFn>
+constexpr DOP DOP::compute(Iterator vertices_begin, Iterator vertices_end, UnaryFn&& fn_get_position) noexcept {
+    return std::accumulate(vertices_begin, vertices_end, DOP{}, [fn_get_position = std::forward<UnaryFn>(fn_get_position)](DOP bounds, const auto& element) {
+        const auto vertex = fn_get_position(element);
+        return bounds.unite_with(vertex);
+    });
+}
+
+inline DOP DOP::transformed(const math::float4x4& transform) const noexcept {
+    auto&& points = get_points();
+    return compute(points.begin(), points.end(), [&transform](const auto& pos) {
+        const auto transformed_position = math::mul(transform, pos);
+        return xyz(transformed_position) / transformed_position.w;
+     });
+}
+
+constexpr DOP DOP::unite_with(const DOP& rhv) const noexcept {
+    DOP result = *this;
+    for (size_t i = 0; i < get_num_planes(); ++i) {
+        result.distances_[i] = std::max(distances_[i], rhv.distances_[i]);
+    }
+    return result;
+}
+
+constexpr DOP DOP::unite_with(const math::float3& point) const noexcept {
+    DOP result = *this;
+    result.expand(point);
+    return result;
+}
+
+constexpr void DOP::expand(const math::float3& point) noexcept {
+    for (size_t i = 0; i < get_num_planes(); ++i) {
+        const auto dot = math::dot(point, get_plane_normal(i));
```

Comment:
```suggestion
        const auto projection = math::dot(point, get_plane_normal(i));
```
`dot` doesn't describe your intent. Feel free to change it to something more verbose if you want.